src/qutip_cupy/cupy_dense.py

```diff
@@ -0,0 +1,36 @@
+import cupy as cp
+import numbers
+from qutip.core import data
+class CuPyDense(data.Data):
+    def __init__(self, data, shape=None, copy=True):
+        base = cp.array(data, dtype=cp.complex128, order='K', copy=copy)
+        if shape is None:
```

Comment:
Hopefully a lot of this shape handling code can be removed once https://github.com/qutip/qutip/pull/1584 is merged.

Reply:
No, all the shape handling will still be required - it's up to base classes to ensure they're passing sensible shapes, because only they know what they should default to given their data.

Reply:
Good point, but hopefully the property can go away and we can just call `super().__init__(shape)` instead.

Reply:
Yeah, the extra property won't be necessary any more for sure.

Reply:
It is removed as an attribute, but kept as an argument. I somewhat changed the if-ifelse shape checks so that reshape is  only enforced when both possible and needed.